cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,303 @@
+		Src: []string{
+			metal.ProvisioningEventPlannedReboot.String(),
+		},
+		Dst: metal.ProvisioningEventPlannedReboot.String(),
+	},
+	{
+		Name: metal.ProvisioningEventPlannedReboot.String(),
+		Src: []string{
+			metal.ProvisioningEventPXEBooting.String(),
+			metal.ProvisioningEventPreparing.String(),
+			metal.ProvisioningEventRegistering.String(),
+			metal.ProvisioningEventWaiting.String(),
+			metal.ProvisioningEventInstalling.String(),
+			metal.ProvisioningEventBootingNewKernel.String(),
+			metal.ProvisioningEventPhonedHome.String(),
+		},
+		Dst: metal.ProvisioningEventPlannedReboot.String(),
+	},
+	{
+		Name: metal.ProvisioningEventMachineReclaim.String(),
+		Src: []string{
+			metal.ProvisioningEventPXEBooting.String(),
+			metal.ProvisioningEventPreparing.String(),
+			metal.ProvisioningEventRegistering.String(),
+			metal.ProvisioningEventWaiting.String(),
+			metal.ProvisioningEventInstalling.String(),
+			metal.ProvisioningEventBootingNewKernel.String(),
+			metal.ProvisioningEventPhonedHome.String(),
+		},
+		Dst: metal.ProvisioningEventMachineReclaim.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventPreparing.String(),
+		},
+		Dst: metal.ProvisioningEventPreparing.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventRegistering.String(),
+		},
+		Dst: metal.ProvisioningEventRegistering.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventWaiting.String(),
+		},
+		Dst: metal.ProvisioningEventWaiting.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventInstalling.String(),
+		},
+		Dst: metal.ProvisioningEventInstalling.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventBootingNewKernel.String(),
+		},
+		Dst: metal.ProvisioningEventBootingNewKernel.String(),
+	},
+}
+
+// every callback expects arguments of types *ProvisioningFSM, *metal.ProvisioningEvent
+var callbacks = fsm.Callbacks{
+	"enter_" + metal.ProvisioningEventRegistering.String():      handleStateTransition,
+	"enter_" + metal.ProvisioningEventWaiting.String():          handleStateTransition,
+	"enter_" + metal.ProvisioningEventInstalling.String():       handleStateTransition,
+	"enter_" + metal.ProvisioningEventBootingNewKernel.String(): handleStateTransition,
+	"enter_" + metal.ProvisioningEventMachineReclaim.String():   handleStateTransition,
+
+	"enter_" + metal.ProvisioningEventPXEBooting.String():    resetFailedReclaim,
+	"enter_" + metal.ProvisioningEventPreparing.String():     resetFailedReclaim,
+	"enter_" + metal.ProvisioningEventPlannedReboot.String(): handlePlannedRebootEvent,
+	"enter_" + metal.ProvisioningEventPhonedHome.String():    handlePhonedHomeEvent,
+}
+
+func handleStateTransition(e *fsm.Event) {
```

Comment:
You can define the callbacks on `ProvisioningFSM` and define the callbacks inline when creating it. Then you don't need to pass it through the arguments.